File: 290. Word Pattern.rb
# @param {String} pattern
# @param {String} str
# @return {Boolean}
def word_pattern(pattern, str)
    str = str.split(' ')
    h1 = {}
    h2 = {}
    return false if pattern.length != str.length
    for i in 0...str.length
        p =pattern[i]
        s = str[i]
        if h1[p].nil? && h2[s].nil? 
            h1[p] = s
            h2[s]=p
        elsif h1[p] == s && h2[s] == p
            next
        else
            return false
        end
    end
    true
end

File: 290. Word Pattern.go
func wordPattern(pattern string, s string) bool {
	words := strings.Split(s, " ")
	if len(words) != len(pattern) {
		return false
	}

	m := make(map[byte]string)
	m2 := make(map[string]bool)
	for i := 0; i < len(pattern); i++ {
		if m[pattern[i]] == "" && !m2[words[i]] {
			m[pattern[i]] = words[i]
            m2[words[i]] = true
		} else if m[pattern[i]] != words[i] {
			return false
		}
	}
	return true
}